Fixes #3293 add FactoryGirl to tests

GregSutcliffe opened the pull request:
Also adds an example factory.

GregSutcliffe committed:
Fixes #3293 add FactoryGirl to tests

domcleal commented:
Test failures:

<pre>
Installing factory_girl (4.2.0) 
Gem::InstallError: factory_girl requires Ruby version >= 1.9.2.
</pre>

I'd rather not have an unused example there, is there something (like settings) that's easy to convert?

GregSutcliffe commented:
@domcleal pinned to Factorygirl 2.X for 1.8 compat. Added a real example involved migrated Usergroups - Settings is a bit of a special case as deleting the fixtures causing interesting effects with the initializers :P

GregSutcliffe commented:
It should also be noted that "example" is a bit misleading - the code in factories/settings.rb is a valid Setting factory and is used by foreman_default_hostgroup to build a Setting without needing to define it's own Factory.

domcleal commented on bundler.d/test.rb:
```diff
@@ -14,4 +14,5 @@
   gem 'database_cleaner', '0.9.1'
   gem 'launchy'
   gem 'spork'
+  gem 'factory_girl_rails', '~> 1.2', :require => false
```
1.x or 2.x?

GregSutcliffe replied:
This is correct, fg_rails 1 pulls in fg 2

GregSutcliffe replied:
https://github.com/thoughtbot/factory_girl/blob/v2.6.5/GETTING_STARTED.md#update-your-gemfile

domcleal replied:
Gotcha, thanks.

domcleal commented:
You're going to hate this, but the integration tests failed :)

Noted about the example, I see, thanks.

dmitri-d commented:
More of a nitpick, but it's better to use sequences and generators instead of hard-coded values (you end up with better tests). https://github.com/thoughtbot/factory_girl/blob/master/GETTING_STARTED.md#lazy-attributes has a good example.

dmitri-d commented on test/unit/usergroup_test.rb:
```diff
@@ -130,29 +132,29 @@ def setup_user operation
   test "user with destroy permissions should be able to destroy" do
+    record = FactoryGirl.create(:usergroup)
     setup_user "destroy"
-    record =  Usergroup.first
```
Edit: NM!

domcleal commented:
@witlessbird so random strings for things like names and descriptions?

dmitri-d commented:
@domcleal: I usually try to automatically generate as many fields as possible, but yours are most typical examples.

domcleal commented:
I like the idea, what do you think @GregSutcliffe?  The only place it matters is one integration test, where we could instead specify the name we care about in the create call.  (Just thinking to get us into "best practices" from the start.)

Otherwise looking good I think.

GregSutcliffe commented:
Trivial to add if you want it as a first example.

domcleal commented:
I think it'd be good if it gets us using it well, as people will only start copying your first examples :)

GregSutcliffe commented:
Copying from me is generally a bad idea.... :) I'll get that added now.

GregSutcliffe commented:
Sadly `generate(:random_string)` appears to be a FG4 thing, but I've added some sequences and converted some of the basic Usergroup model tests to call FG.build where appropriate for in-memory tests. Looks nice actually :)

@domcleal @witlessbird how does that look to you for an example now?

dmitri-d commented:
@GregSutcliffe: looks good! (you may know this, but you can re-use sequences in multiple factories, but only if you define them outside of "factory" scope. This mostly compensates lacks of 'generate').

GregSutcliffe commented:
Ah yes, I knew you could re-use them, but hadn't made the jump to a generate replacement. Maybe later, I cba altering it again :)

domcleal commented:
Thanks @GregSutcliffe, merged as 3035495 for Foreman 1.3.1 as requested.
